lemmatization/services/chinese.py

```diff
@@ -0,0 +1,223 @@
+
+    def prepare_for_segmentation(self, text: str) -> str:
+        """
+        Perform any processing steps required to optimize string of text for LAC segmentation.
+        """
+        # LAC seems to work better with just \n delimiters, so remove any \r (e.g. in \r\n sequences)
+        return text.replace("\r", "")
+
+    def tokenize_chinese_text(self, text: str) -> Iterable[str]:
+        return self.lac.run(text)
+
+    def re_tokenize(self, tokens: Iterable[str]) -> Iterable[str]:
+        """
+        Ensure any tokens containing pipe hints are split, and that newlines are also split into
+        their own tokens, in order to help with lattice lookups downstream. (LAC sometimes includes
+        newlines in lemma tokens instead of splitting them out.)
+
+        >>> re_tokenize(["a|b", "c\nd"])
+        ["a", "b", "c", "\n", "d"]
+        """
+        for token in tokens:
+            # collect all characters in the token but split on the pipe character and newlines
+            subtoken = ""
+            for char in token:
+                # consider pipes and newlines to be token breaks; in practice, LAC doesn't always do this
+                if char in ("|", "\n"):
+                    if subtoken:
+                        yield subtoken
+                        subtoken = ""
+                    # if it is a newline, emit it as a separate token; if a pipe, do not emit it
+                    if char == "\n":
+                        yield char
+                else:
+                    subtoken += char
+            if subtoken:
+                yield subtoken
+
+
+# this captures a wide swath of Chinese characters and related symbols, such as
+# the Chinese "zero" (〇 or \u3007) and middle dot (· or \u00B7, used with names)
+# without capturing common punctuation (e.g. in the \u3000-303F range)
+# cf https://en.wikipedia.org/wiki/CJK_Unified_Ideographs
+CHINESE_RE = re.compile(
+    r"^[\u00B7-\u00B7\u3007-\u3007\u3400-\u4DBF\u4E00-\u9FEF\U00020000-\U0002EBFF]+$"
+)
+
+
+def normalize_chinese(text: str) -> str:
+    """
+    Performs Unicode normalization on text. cf Unicode Standard section 2.12.
+    Uses NFC (Normalization Form C). Also selectively converts full-width
+    pipe (｜ \uFF5C) and underscore (＿ \uFF3F) to half-width (ASCII) | \u007C
+    and _ \u005F (so they can be used to give token boundary hints).
+
+    >>> normalize_chinese("1｜2＿3")
+    "1|2_3"
+
+    There are some issues with using NFKC, e.g.:
+    - common full-width forms like Chinese commas are transformed into half-width using NFKC
+    (and other full-width forms like the pipe/vertical bar are not transformed).
+    """
+    normalized_text = unicodedata.normalize("NFC", text)  # normalize
+    return normalized_text.replace("\uFF5C", "|").replace("\uFF3F", "_")
+
+
+def normalize_token(token: str) -> str:
+    """
+    Perform transformations or adjustments to individual Chinese tokens
+    """
+    # currently a no-op
+    return token
```

Comment:
What do you consider the use of defining a function that only returns its argument? Just curious.

Reply:
Hmm, yes, a good question. I defined this function to provide a place for any future normalization to happen, and as an indication that the second part of the "triple" is meant to be a normalized form (is it documented anywhere?). Also the process evolved by following the pattern of the other lemmatizers. But now that I look at it, there seem to be two approaches that would be better: it could simply be gone to reduce confusion, and returning `(token, token, following)` can be documented to support that choice; or, if we left it, it makes more sense for it to be a class-specific method, as it would likely vary with any given implementation. 

In the end when looking at it I opted with c791be7 to make everything encapsulated, either in the base class or the LAC subclass. I hope it's easier to understand that way, and supports easier alternative subclassing, but it might be overkill in some people's eyes.